File: test/src/tc/es/junit/src/org/iotivity/service/es/test/btc/ESEasySetupTest.java
/******************************************************************
*
* Copyright 2016 Samsung Electronics All Rights Reserved.
*
*
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
******************************************************************/

package org.iotivity.service.es.test.btc;

import org.iotivity.service.es.test.helper.ESEnrolleeHelper;
import org.iotivity.service.easysetup.mediator.ESException;
import org.iotivity.service.easysetup.mediator.EasySetup;
import org.iotivity.base.OcResource;
import org.iotivity.base.OcPlatform;

import static org.iotivity.service.es.test.helper.ESUtility.*;

import android.test.AndroidTestCase;

public class ESEasySetupTest extends AndroidTestCase {

    private OcResource          ocResource;
    private ESEnrolleeHelper    eSEnrolleeHelper;
    private static final String RESOURCE_NAME          = "/a/light";
    private static final String PROV_RESOURCE_TYPE     = "ocf.wk.prov";
    private static final String INVALID_RESOURCE_TYPE  = "oic.es.invalid";
    private static final String INVALID_INTERFACE_TYPE = OcPlatform.DEFAULT_INTERFACE;
    private static final String BATCH_INTERFACE_TYPE   = OcPlatform.BATCH_INTERFACE;

    protected void setUp() throws Exception {
        super.setUp();
        eSEnrolleeHelper = new ESEnrolleeHelper();
    }

    protected void tearDown() throws Exception {
        super.tearDown();
        eSEnrolleeHelper = null;
        ocResource = null;
    }

    /**
     * @since 2016-08-14
     * @see none
     * @objective test getInstance API positively
     * @target public synchronized static EasySetup getInstance(Context context)
     * @test_data context
     * @pre_condition none
     * @procedure 1. Call getInstance API with application context. 2. Check
     *            easy setup intsance is not null.
     * @post_condition none
     * @expected initialization of the easy setup
     */
    public void testESGetInstance_SRC_RSV_P() {
        assertNotNull("Fail to cretae easy setup instance",
                EasySetup.getInstance(getContext()));
    }

    /**
     * @since 2016-08-14
     * @see none
     * @objective test getInstance API negatively
     * @target public synchronized static EasySetup getInstance(Context context)
     * @test_data null context
     * @pre_condition none
     * @procedure 1. Call getInstance API with null context.
     * @post_condition none
     * @expected exception thrown.
     */
    public void testESGetInstanceWithNullContext_ETC_NV_N() {
        try {
            EasySetup.getInstance(null);
            fail(EXCEPTION_SHOULD_BE_THROWN);
        } catch (Exception e) {
            assertTrue(ES_EXCEPTION_SHOULD_BE_THROWN, e instanceof ESException);
        }
    }

    /**
     * @since 2016-08-14
     * @see none
     * @objective test createRemoteEnrollee API can create remote enrolle.
     * @target public synchronized RemoteEnrollee
     *         createRemoteEnrollee(OcResource enrolleeResource)
     * @test_data ocresource
     * @pre_condition create oc resource
     * @procedure 1. Call createRemoteEnrollee API with created resource. 2.
     *            Check enrollee object is not null.
     * @post_condition none
     * @expected Successful creation of remote enrollee object
     */
    public void testESCreateRemoteEnrollee_P_RSV() {
        ocResource = eSEnrolleeHelper.createEnrolleeResource(getContext(),
                RESOURCE_NAME, PROV_RESOURCE_TYPE, BATCH_INTERFACE_TYPE);
        assertNotNull("Resource object should not null", ocResource);
        assertNotNull("RemoteEnrollee should not null", EasySetup
                .getInstance(getContext()).createRemoteEnrollee(ocResource));
    }

    /**
     * @since 2016-08-18
     * @see none
     * @objective test createRemoteEnrollee API with resource containing invalid
     *            interface
     * @target public synchronized RemoteEnrollee
     *         createRemoteEnrollee(OcResource enrolleeResource)
     * @test_data context
     * @pre_condition none
     * @procedure 1. Call createRemoteEnrollee API with resource containing
     *            invalid resource type. 2. Check remote enrollee object is null
     * @post_condition none
     * @expected Null remote enrollee will be returned.
     */
    public void testESCreateRemoteEnrolleeWithInvalidInterfaceType_USV_SRC_N() {
        ocResource = eSEnrolleeHelper.createEnrolleeResource(getContext(),
                RESOURCE_NAME, INVALID_RESOURCE_TYPE, BATCH_INTERFACE_TYPE);
        assertNotNull("Resource object should not null", ocResource);
        assertNull("RemoteEnrollee should null", EasySetup
                .getInstance(getContext()).createRemoteEnrollee(ocResource));
    }

    /**
     * @since 2016-08-18
     * @see none
     * @objective test createRemoteEnrollee API with resource containing invalid
     *            interface
     * @target public synchronized RemoteEnrollee
     *         createRemoteEnrollee(OcResource enrolleeResource)
     * @test_data context
     * @pre_condition none
     * @procedure 1. Call createRemoteEnrollee API with resource containing
     *            invalid interface. 2. Check remote enrollee object is null.
     * @post_condition none
     * @expected Null remote enrollee will be returned.
     */
    public void testESCreateRemoteEnrolleeWithInvalidResourceType_USV_SRC_N() {
        ocResource = eSEnrolleeHelper.createEnrolleeResource(getContext(),
                RESOURCE_NAME, PROV_RESOURCE_TYPE, INVALID_INTERFACE_TYPE);
        assertNotNull("Resource object should not null", ocResource);
        assertNull("RemoteEnrollee should null", EasySetup
                .getInstance(getContext()).createRemoteEnrollee(ocResource));
    }

    /**
     * @since 2016-08-18
     * @see none
     * @objective test createRemoteEnrollee API with resource containing invalid
     *            interface
     * @target public synchronized RemoteEnrollee
     *         createRemoteEnrollee(OcResource enrolleeResource)
     * @test_data context
     * @pre_condition none
     * @procedure 1. Call createRemoteEnrollee API with null resource
     * @post_condition none
     * @expected should throw exception
     */
    public void testESCreateRemoteEnrolleeWithNullResource_ETC_N() {
        ocResource = null;
        try {
            EasySetup.getInstance(getContext())
                    .createRemoteEnrollee(ocResource);
            fail(EXCEPTION_SHOULD_BE_THROWN);
        } catch (Exception e) {
            assertTrue(ES_EXCEPTION_SHOULD_BE_THROWN, e instanceof ESException);
        }
    }
}

File: test/src/tc/es/gtest/src/btc/mediator/ESEasySetupTest.cpp
//******************************************************************
//
// Copyright 2015 Samsung Electronics All Rights Reserved.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#include <gtest/gtest.h>
#include "CommonUtil.h"
#include <atomic>
#include <functional>
#include <condition_variable>
#include <mutex>
#include <chrono>
#include <exception>
#include "easysetup.h"
#include <escommon.h>
#include "ESEnrolleeCommon.h"
#include "ESMediatorHelper.h"
#include <iostream>

using namespace std;
using namespace OC;
using namespace OIC::Service;

std::mutex g_mutexForCondition;
std::shared_ptr< RemoteEnrollee > g_remoteEnrollee;
std::condition_variable g_responseCon;
std::chrono::milliseconds g_waitForResponse(1000);

class ESEasySetupTest_btc: public ::testing::Test
{
public:
    std::shared_ptr< OC::OCResource > m_enrolleeResource;
    static std::shared_ptr< RemoteEnrollee > s_remoteEnrollee;

    std::shared_ptr<OC::OCResource> CreateNotProvResource()
    {
        OCConnectivityType connectivityType = CT_DEFAULT;
        std::vector<std::string> types = {"oic.wk.notprov"};
        std::vector<std::string> ifaces = {DEFAULT_INTERFACE};

        return OCPlatform::constructResourceObject("coap://192.168.1.2:5000",
                                                   "/NotProvisioningResURI",
                                                   connectivityType,
                                                   false,
                                                   types,
                                                   ifaces);
    }

protected:

    virtual void SetUp()
    {
        CommonUtil::runCommonTCSetUpPart();
    }

    virtual void TearDown()
    {
        g_remoteEnrollee = nullptr;
        CommonUtil::runCommonTCTearDownPart();
    }

};

/**
 * @since 2016-08-16
 * @objective Test 'createRemoteEnrollee' function with positive basic way
 * @target Void createRemoteEnrollee();
 * @test_data	1. OCResource::Ptr constructResourceObject()
 * 				2. discoverRemoteEnrolleeCb callback
 * @pre_condition None
 * @procedure Perform createRemoteEnrollee
 * @post_condition None
 * @expected creating a remote Enrollee instance
 **/
TEST_F(ESEasySetupTest_btc, CreateRemoteEnrollee_SRC_P)
{
    try
    {
        ESMediatorHelper esMediatorHelper;
        esMediatorHelper.findEnrolleeResource();
        
        g_remoteEnrollee = EasySetup::getInstance()->createRemoteEnrollee(
                ESMediatorHelper::s_curResource);

        if(g_remoteEnrollee==NULL)
        {
            SET_FAILURE("Remote enrollee was not created.");
        }
    }
    catch (exception& e)
    {
        SET_FAILURE("Exception occurred in get ProvStatus: " + std::string(e.what()));
    }
}

/**
 * @since 2016-08-16
 * @objective Test 'createRemoteEnrollee' function in negative way
 * @target Void createRemoteEnrollee();
 * @test_data   1. OCResource::Ptr constructResourceObject()
 *              2. Null parameter
 * @pre_condition None
 * @procedure Perform createRemoteEnrollee
 * @post_condition None
 * @expected return null pointer
 **/
TEST_F(ESEasySetupTest_btc, CreateRemoteEnrolleeWithNonProvResource_USV_N)
{
    auto s_remoteEnrollee = EasySetup::getInstance()->createRemoteEnrollee(CreateNotProvResource());
    EXPECT_EQ(nullptr, s_remoteEnrollee)<<"Failed for Not creating resource.";

    g_remoteEnrollee = EasySetup::getInstance()->createRemoteEnrollee(NULL);
    EXPECT_EQ(nullptr, g_remoteEnrollee)<<"Failed for null parameter.";
}

